<a href="https://colab.research.google.com/github/lwy2107/GK_co_work/blob/master/%EC%8B%A0%EA%B2%BD%EB%A7%9D_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from tensorflow import keras
(train_input, train_target), (test_input, test_target)=\
keras.datasets.fashion_mnist.load_data()

4422102/4422102 [==============================] - 1s 0us/step


In [3]:
from sklearn.model_selection import train_test_split
train_scaled=train_input/255.0
train_scaled=train_scaled.reshape(-1,28*28)
train_scaled, val_scaled, train_target, val_target=train_test_split(
    train_scaled, train_target, test_size=0.2, random_state=42)

In [4]:
dense1=keras.layers.Dense(100, activation='sigmoid', input_shape=(784,))#시그모이드 함수를 사용 신경망은 첫번째층에 무조건 input_shape매개변수크기를 입력해야함
dense2=keras.layers.Dense(10, activation='softmax') # 소프트맥스 층 선언

In [5]:
model = keras.Sequential([dense1, dense2]) #심층 신경망 사용 첫째층은 시그모이드 두번째층은 소프트맥스층을 사용하였다.

In [6]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               78500     
                                                                 
 dense_1 (Dense)             (None, 10)                1010      
                                                                 
Total params: 79510 (310.59 KB)
Trainable params: 79510 (310.59 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


지금 summury메서드를 이용하여 출력을 해보았다.
모델은 sequential이라는 이름을 사용하고 있고 층은 총 2개의 층으로 구성된 딥러닝형식으로 구성되어있고 출력형식은 첫번째 층에서는 데이터셋의 데이터를를 입력하여 100개의 값을 출력하도록 하고 그 100개의 출력된값을 다시 두번째층에서 계산하여 10개의 결과값을 내도록하였다. 이를 파라미터로 나타냈다. 첫번째층에서는 784픽셀의 데이터에 100개의 뉴런 100개의 절편을 계산하게되는데
784*100+100 으로구성되어있다. 즉 78500개라는 파라미터 값이 나오게된다.
그다음층에서는 100개의 계산된 결과값을 받았고 10개의 뉴런의 값을 출력하게 하고 그에 맞는 절편을 구성하게된다.
100*10+10로 구성된다. 그 결과 파라미터값은 1010으로 나오게 되고 모든값을 더하게되면 79510개의 파라미터 값이 나오게된다.

In [7]:
model = keras.Sequential([
    keras.layers.Dense(100,activation='sigmoid', input_shape=(784,),name='hidden'),
    keras.layers.Dense(10, activation='softmax', name='output')], name='패션MNIST 모델')
model.summary()

Model: "패션MNIST 모델"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 hidden (Dense)              (None, 100)               78500     
                                                                 
 output (Dense)              (None, 10)                1010      
                                                                 
Total params: 79510 (310.59 KB)
Trainable params: 79510 (310.59 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [8]:
model = keras.Sequential()
model.add(keras.layers.Dense(100, activation='sigmoid', input_shape=(784,)))
model.add(keras.layers.Dense(10, activation='softmax'))
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_2 (Dense)             (None, 100)               78500     
                                                                 
 dense_3 (Dense)             (None, 10)                1010      
                                                                 
Total params: 79510 (310.59 KB)
Trainable params: 79510 (310.59 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [9]:
model.compile(loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.5706 - accuracy: 0.8062
Epoch 2/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.4127 - accuracy: 0.8525
Epoch 3/5
1500/1500 [==============================] - 5s 4ms/step - loss: 0.3765 - accuracy: 0.8641
Epoch 4/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3542 - accuracy: 0.8718
Epoch 5/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3356 - accuracy: 0.8785


위의 신경망을 에포크 5회로 훈련시켜보았다.
에포크가 가면갈수록 예측의 정확도가 올라가는것을 알수 있다.

현재 사용중인 선형함수들은 단점이 있다. 0~1이라는 소수점을 모두 계산해야하다보니 정확한 값을 내기가 힘들어진다. 그러므로 렐루함수를 사용해서 0or1로 하여 결과를 나타내게 하면 보다 효과적으로 결과값을 도출해낼수 있다. 그리고 사용량도 줄어들게 된다.

In [10]:
model = keras.Sequential()
model.add(keras.layers.Flatten(input_shape=(28,28)))
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(10,activation='softmax'))
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense_4 (Dense)             (None, 100)               78500     
                                                                 
 dense_5 (Dense)             (None, 10)                1010      
                                                                 
Total params: 79510 (310.59 KB)
Trainable params: 79510 (310.59 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [18]:
(train_input, train_target),(test_input, test_target)=keras.datasets.fashion_mnist.load_data()
train_scaled = train_input/255.0
train_scaled, val_scaled, train_target, val_target = train_test_split(
    train_scaled, train_target, test_size=0.2, random_state=42
)


In [19]:
model.compile(loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 6s 4ms/step - loss: 0.5356 - accuracy: 0.8102
Epoch 2/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3898 - accuracy: 0.8587
Epoch 3/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3544 - accuracy: 0.8736
Epoch 4/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3312 - accuracy: 0.8807
Epoch 5/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3158 - accuracy: 0.8876


In [20]:
model.evaluate(val_scaled, val_target)

375/375 [==============================] - 1s 2ms/step - loss: 0.3655 - accuracy: 0.8737


[0.3654516935348511, 0.8737499713897705]

케라스는 기본적으로 하강기법을 이용하여 학습을 하는데 이 하강기법들을 옵티마이저라 부른다. 미니배치(데이터를 작은묶음으로 계산)하여 예측값을 낸다.

In [21]:
##model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics='accuracy')
#sgd=keras.optimizers.SGD(learning_rate=0.1)#학습률을 0.1 즉10%로 설정하고싶을때 이렇게 넣으면된다.
#model.compile(optimizer=sgd, loss='sparse_categorical_crossentropy', metrics='accuracy') 위의 sgd와 동일한 옵티마이져이다. 원래 이렇게 객채선언도 해야한다.
#하지만 번거로움을 피하고자 'sgd'로묶어 바로 사용하도록 하였다.

In [ ]:
##dgf=keras.optimizers.SGD(momentum=0.9, nesterov=True) #모멘텀이 0.9이며 네스테로프가속경사를 사용하는 예이다.

네스테로프가속경사는 모멘텀을 2회반복하는 것이다. 기존의 경사하강보다 훨씬빠르게 최적점에 다다를수 있게된다. 이렇게 하여 학습하다보면 학습률을 줄여야하는데 이렇게 늘리고 줄이는것을 적응형학습이라 한다.
이런방법으로 튜닝하여 더 나은 예측을 얻을수 있다.
아래에는 적응형학습률을 사용하는 옵티마이저이다.

In [26]:
adagrad=keras.optimizers.Adagrad()
model.compile(optimizer=adagrad, loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)
rmsprop=keras.optimizers.RMSprop()
model.compile(optimizer=rmsprop, loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2585 - accuracy: 0.9054
Epoch 2/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2541 - accuracy: 0.9066
Epoch 3/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2523 - accuracy: 0.9075
Epoch 4/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2510 - accuracy: 0.9083
Epoch 5/5
1500/1500 [==============================] - 7s 5ms/step - loss: 0.2502 - accuracy: 0.9086
Epoch 1/5
1500/1500 [==============================] - 6s 4ms/step - loss: 0.2903 - accuracy: 0.8922
Epoch 2/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2829 - accuracy: 0.8963
Epoch 3/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2751 - accuracy: 0.9000
Epoch 4/5
1500/1500 [==============================] - 6s 4ms/step - loss: 0.2685 - accuracy: 0.9031
Epoch 5/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2624 - accurac

이러한 적응형학습법이 있다. 이와 더불이 가장 쉽게 사용할수 있는 Adam을 사용하여 실습해본다.

In [24]:
model=keras.Sequential()
model.add(keras.layers.Flatten(input_shape=(28,28)))
model.add(keras.layers.Dense(100,activation='relu'))
model.add(keras.layers.Dense(10,activation='softmax'))

In [25]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 6s 3ms/step - loss: 0.5172 - accuracy: 0.8191
Epoch 2/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3895 - accuracy: 0.8593
Epoch 3/5
1500/1500 [==============================] - 6s 4ms/step - loss: 0.3463 - accuracy: 0.8742
Epoch 4/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3212 - accuracy: 0.8830
Epoch 5/5
1500/1500 [==============================] - 6s 4ms/step - loss: 0.3029 - accuracy: 0.8885


In [27]:
model.evaluate(val_scaled, val_target)

375/375 [==============================] - 1s 2ms/step - loss: 0.3662 - accuracy: 0.8827


[0.36623021960258484, 0.8826666474342346]